File: src/hackerrank/all/JumpingOnClouds.java
package src.hackerrank.all;

public class JumpingOnClouds {

    public static void main(String[] args) {
        int jumps = jumpingOnClouds(new int[]{0, 0, 1, 0, 0, 1, 0});
        int jumps2 = jumpingOnClouds(new int[]{0, 0, 0, 0, 1, 0});
        int jumps3 = jumpingOnClouds(new int[]{0, 0, 0, 1, 0, 0});
        return;
    }

    static int jumpingOnClouds(int[] c) {
        int jumps = 0;

        for (int i = 0; i < c.length;) {
            if (isPossibleToJump(c, i + 2)) {
                i+=2;
                jumps++;
            } else if (isPossibleToJump(c, i + 1)) {
                i++;
                jumps++;
            } else {
                break;
            }
        }

        return jumps;
    }

    static boolean isPossibleToJump(int[] c, int pos) {
        if (pos >= c.length) return false;
        if (c[pos] == 1) return false;

        return true;
    }
}

// https://www.hackerrank.com/challenges/jumping-on-the-clouds/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=warmup

File: src/golang/arrays/jumping_on_clouds.go
package main

import "fmt"

/** https://www.hackerrank.com/challenges/jumping-on-the-clouds/problem?h_l=interview&playlist_slugs%5B%5D=interview-preparation-kit&playlist_slugs%5B%5D=warmup */

func main() {
	fmt.Println(jumpingOnClouds([]int32{0, 0, 1, 0, 0, 1, 0}))
}

func jumpingOnClouds(c []int32) int32 {
	currentIndex := 0
	jumps := 0

	for {
		if canMakeDoubleStep(currentIndex, c) {
			currentIndex += 2
		} else {
			currentIndex++
		}

		jumps++

		if currentIndex == len(c)-1 {
			return int32(jumps)
		}
	}
}

func canMakeDoubleStep(currentIndex int, c []int32) bool {
	if currentIndex+2 <= len(c)-1 && c[currentIndex+2] != 1 {
		return true
	} else {
		return false
	}
}
